### ISSUE TITLE
DevContainer: Changing UID/GID for remote user fails when GID exists (Linux/WSL)

### ISSUE THREAD

**Issue by MartinKolbAtWork** [microsoft/vscode-remote-release#7284]
- VSCode Version: 1.72.2
- Local OS Version: Windows 21H2
- Remote OS Version: WSL/Ubuntu and Suse Linux Enterprise 15 SP3
- Remote Extension/Connection Type: WSL/SSH

I am trying to use the feature that specifying `remoteUser` in `devcontainer.json`, the UID and GID of the remoteUser is changed to match the UID/GID of the local user.
In my example, the local user has UID 1002 and the GID 100:

```
$ id
uid=1002(martin) gid=100(users) groups=100(users),476(docker),1002(developers)
```

In the container, the user has the UID 1000 and GID 1000, and the group 100 does also exist:

```
$ id
uid=1000(vscode) gid=1000(vscode) groups=1000(vscode)
```

When the container is created by VS Code, the Docker build for changing the UID/GID is started with parameters “NEW_UID=1002” and “NEW_GID=100”, which is exactly specifying the correct values of the host user:

`Start: Run: docker build -f /tmp/vsch-martin/updateUID.Dockerfile-0.251.0 -t vsc-dc-bc3f93362ac8811fa34f7aa3e422f51b-uid --build-arg BASE_IMAGE=vsc-dc-bc3f93362ac8811fa34f7aa3e422f51b --build-arg REMOTE_USER=vscode --build-arg NEW_UID=1002 --build-arg NEW_GID=100 --build-arg IMAGE_USER=root /tmp/vsch-martin`

However, in the Dockerfile, a check is performed, if the “NEW_GID” does already exist. And if it exists, no changes are performed:

```
ARG BASE_IMAGE
FROM $BASE_IMAGE

USER root

ARG REMOTE_USER
ARG NEW_UID
ARG NEW_GID
SHELL ["/bin/sh", "-c"]
RUN eval $(sed -n "s/${REMOTE_USER}:[^:]*:\([^:]*\):\([^:]*\):[^:]*:\([^:]*\).*/OLD_UID=\1;OLD_GID=\2;HOME_FOLDER=\3/p" /etc/passwd); \
	eval $(sed -n "s/\([^:]*\):[^:]*:${NEW_UID}:.*/EXISTING_USER=\1/p" /etc/passwd); \
	eval $(sed -n "s/\([^:]*\):[^:]*:${NEW_GID}:.*/EXISTING_GROUP=\1/p" /etc/group); \
	if [ -z "$OLD_UID" ]; then \
		echo "Remote user not found in /etc/passwd ($REMOTE_USER)."; \
	elif [ "$OLD_UID" = "$NEW_UID" -a "$OLD_GID" = "$NEW_GID" ]; then \
		echo "UIDs and GIDs are the same ($NEW_UID:$NEW_GID)."; \
	elif [ "$OLD_UID" != "$NEW_UID" -a -n "$EXISTING_USER" ]; then \
		echo "User with UID exists ($EXISTING_USER=$NEW_UID)."; \
	elif [ "$OLD_GID" != "$NEW_GID" -a -n "$EXISTING_GROUP" ]; then \
		echo "Group with GID exists ($EXISTING_GROUP=$NEW_GID)."; \
	else \
		echo "Updating UID:GID from $OLD_UID:$OLD_GID to $NEW_UID:$NEW_GID."; \
		sed -i -e "s/\(${REMOTE_USER}:[^:]*:\)[^:]*:[^:]*/\1${NEW_UID}:${NEW_GID}/" /etc/passwd; \
		if [ "$OLD_GID" != "$NEW_GID" ]; then \
			sed -i -e "s/\([^:]*:[^:]*:\)${OLD_GID}:/\1${NEW_GID}:/" /etc/group; \
		fi; \
		chown -R $NEW_UID:$NEW_GID $HOME_FOLDER; \
	fi;

ARG IMAGE_USER
USER $IMAGE_USER
```
The script does the check `elif [ "$OLD_GID" != "$NEW_GID" -a -n "$EXISTING_GROUP" ]` and then exits with the message `Group with GID exists ($EXISTING_GROUP=$NEW_GID).`
This means, my user is not adapted at all.
However, this check is IMHO wrong and should be done with the exact opposite semantics. The GID that the user is assigned to must exist, otherwise changing the GID would not work.

Could you please check this, or give me some advise how I can work around it and get my scenario working?

Thanks,
Martin

**Comment by chrmarti** [comment]
The script has no way of checking if the existing container's group with GID 100 is suitable as the group for the user. Due to that we check if the GID is already in use.

**Comment by MartinKolbAtWork** [comment]
@chrmarti Thanks for your reply.

Sorry for my probably dumb question, but under which circumstances will then the automatic UID/GID replacement work?

In my project I have a Dockerfile that uses the rust container as base and creates a user called "vscode" that I shall be used as the user for vscode.
**Dockerfile**:
```
FROM rust:latest
RUN useradd -m vscode
```

The `devcontainer.json` uses this new user "vscode" as "remoteUser" and "containerUser"
**devcontainer.json**:
```
{
    "name": "Rust Dev",
    "build": {
        "dockerfile": "Dockerfile"
    },
    // Needed for ptrace-based debuggers like C++, Go, and Rust
    "runArgs": [
        "--cap-add=SYS_PTRACE",
        "--security-opt",
        "seccomp=unconfined"
    ],
    "remoteUser": "vscode",
    "containerUser": "vscode",
    "updateRemoteUserUID": true
}
```

As mentioned initially, on my Linux system, the user has the UID of 1002 (and other team members in my dev team have other UIDs above 1000) and the default GID of 100, which is a common case to have user GIDs of 1000 and above, which are assigned to the 100 group.

Can you briefly explain how I need to adapt the Dockerfile and devcontainer.json to make the automatic UID/GID adaptation of VS Code work?

**Comment by chrmarti** [comment]
Got it. One thing you could try in your Dockerfile is to remove the `users` group from `/etc/group`, that should allow the UID/GID update to run.

We could improve the updater script to handle the case where the target group exists and is named `users`.

**Comment by MartinKolbAtWork** [comment]
@chrmarti , thanks for the suggested workaround. I deleted the 'users' group in the container, and then it worked in my scenario.
However, the preparation of the container should be agnostic to the concrete environment for the developers using a project with '.devcontainer'.
So, I'd appreciate very much if the behavior of the extension can be improved to handle assignment of host users to standard system groups (like 'users') gracefully.

**Comment by davendu** [comment]
Sorry to bother here. I upgraded `ms-vscode-remote.remote-containers` to `v0.311.0`, and it seems using `dev-containers-cli-0.311.0`:

```
[6049 ms] Start: Run: docker build -f /tmp/devcontainercli-davendu/updateUID.Dockerfile-0.51.2 -t vsc-...-uid --build-arg BASE_IMAGE=mcr.microsoft.com/devcontainers/typescript-node:1-20-bullseye --build-arg REMOTE_USER=node --build-arg NEW_UID=1002 --build-arg NEW_GID=100 --build-arg IMAGE_USER=root /tmp/devcontainercli-davendu/empty-folder
```

Is there any way to let it use `updateUID.Dockerfile-0.51.3`?

**Comment by chrmarti** [comment]
@davendu This is now available with Dev Containers 0.313.0-pre-release.

Steps for verification:
- Use Linux as the host either locally or through SSH or a tunnel.
- On the Linux host have a folder with a dev container with a user that has UID and GID that are different from the host's user. (see files below)
- Reopen that folder in the dev container and verify that the user in the container was updated to the host's user UID and GID. (run `id` on the host and in the container to compare)
- Make the container have a group with the same GID as the host's user to test the fix.
- Rebuild the dev container and verify that the UID of the container user was updated, but not the GID.

`.devcontainer/devcontainer.json`:
```json
{
	"build": {
		"dockerfile": "Dockerfile"
	},
	"remoteUser": "foo"
}
```

`.devcontainer/Dockerfile`:
```Dockerfile
FROM debian:latest

# RUN addgroup --gid !HOST_USER_GID! bar || true

RUN addgroup --gid 4321 foo
RUN adduser --uid 1234 --gid 4321 foo
```

(To add the group in the second to last step, uncomment the line in the Dockerfile and replace `!HOST_USER_GID!`.)
